File: frontend/src/types/task.ts
import { Area } from './Area';
import { BaseEntity } from './BaseEntity';
import { TaskStatus } from './Enums';
import { Step } from './Step';
import { Category } from './Category';
import { User } from './User';

export type TaskFilter = {
    areaId?:     string;
    categoryId?: string;
    status?:     TaskStatus;
    active?:     boolean;
    memberId?:   string;
};

export type TaskFilterResponse = {
    tasks: Task[];
    total: number;
};

export type TaskMember = {
    id:     string;
    taskId: string;
    user:   User;
};

export interface Task extends BaseEntity {
    name:           string;
    description?:   string;
    category:       Category;
    area:           Area;
    leader:         User;
    manager:        User;
    status:         TaskStatus;
    active:         boolean;
    steps:          Step[];
    dateLimit?:     string;
    assignedTo:     User[];
    //priority:     TaskPriority;
    //createdBy:    string;
};

File: backend/src/main/java/com/badgerracing/bagder_tasks/domain/entity/Task.java
package com.badgerracing.bagder_tasks.domain.entity;

import com.badgerracing.bagder_tasks.domain.enums.TaskStatus;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.List;

@Entity @Table(name = "tasks")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Task extends BaseEntity {

    @Column(nullable = false)
    private String name;

    private String description;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "area_id", nullable = false)
    private Area area;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "leader_id", nullable = false)
    private User leader;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "manager_id", nullable = false)
    private User manager;

    // status here with take the role of a 'done' boolean variable
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TaskStatus status;

    // a task can be inactive because it's not being currently worked on, but it can be active again,
    // so the 'status' field doesn't need to be linked to the 'active' field.
    // active here is used mainly so that users have a clear feedback on if they should or not
    // worry about the task. so if 'active = false', they can disregard it.
    @Column(nullable = false)
    private boolean active;

    private LocalDateTime dateLimit;

    @OneToMany(
        mappedBy = "task",
        cascade = CascadeType.ALL,
        orphanRemoval = true
    )
    private List<TaskMember> members;

    @OneToMany(
        mappedBy = "task",
        cascade = CascadeType.ALL,
        orphanRemoval = true
    )
    private List<Step> steps;
}